sed/tests/test.bats

```diff
@@ -0,0 +1,22 @@
+source "${BATS_TEST_DIRNAME}/../plan.sh"
+
+pkg_ident=$(grep pkg_ident $BATS_TEST_DIRNAME/../results/last_build.env | cut -d \= -f2)
+
+@test "sed binary exists" {
+ [ -f "/hab/pkgs/$pkg_ident/bin/sed" ]
```

Comment:
This test should execute the binary directly.

Even better, it should use `command` to eliminate any aliasing.

In general this will result in a test that looks something like:

```
@test "Command is on path" {
  [ "$(command -v binary)" ]
}
```

The reason for this is because Habitat guarantees that the package is linked and that path leaking from the operating system cannot happen. In Habitat this is called Binlinking.

I would consider it essential to test that the binary you are calling is exercising the binlink, because otherwise you may have a problem with the `HAB_BINLINK_DIR` environment variable or `pkg_bin_dirs` directory in the plan.

Calling the path directly like this is not always an antipattern, as you may need to do it inside of a plan. In these cases, you can use `pkg_path_for()` to reference the package's binary directory, as habitat's directory may not be where you expect on disk.

This holds true for the rest of the tests here. Call the binary directory instead of using `hab pkg exec` or calling it on disk.

Reply:
> The reason for this is because Habitat guarantees that the package is linked and that path leaking from the operating system cannot happen. In Habitat this is called Binlinking.

I'm a little confused here, the hab subcommand `binlink` (which is a convenience command for providing symbolic links from habitat packages to system paths) is separate from [linking](https://en.wikipedia.org/wiki/Linker_(computing)), which is a build time concern.  Would you mind providing some additional clarification here? 

> I would consider it essential to test that the binary you are calling is exercising the binlink, because otherwise you may have a problem with the `HAB_BINLINK_DIR` environment variable or `pkg_bin_dirs` directory in the plan.

I feel this is actually a bit of an anti-pattern in testing, as it can mask potential issues. For example, if a package relies on a command that is provided by a system path but isn't included in `pkg_deps` you will see a test pass when 'binlinked', but not when executed via `hab pkg exec`.  Example with a simplistic "sed wrapper" package: 

```
[16][default:/src:0]# cat bin/mysed
#!/bin/sh

sed "$@"
[17][default:/src:0]# cat plan.sh
pkg_name="mysed"
pkg_version="0.0.1"
pkg_bin_dirs=(bin)
do_build() { :; }
do_install() {
	cp $PLAN_CONTEXT/bin/mysed $pkg_prefix/bin/mysed
}
```

```
[14][default:/src:0]# hab pkg exec smacfarlane/mysed mysed --help
/hab/pkgs/smacfarlane/mysed/0.0.1/20190410152301/bin/mysed: line 3: sed: command not found
[15][default:/src:127]# /hab/pkgs/smacfarlane/mysed/0.0.1/20190410152301/bin/mysed --help
Usage: sed [OPTION]... {script-only-if-no-other-script} [input-file]...
... elided for space
```

> This holds true for the rest of the tests here. Call the binary directly instead of using `hab pkg exec` or calling it on disk. Habitat's binlink will take care of any PATHing issues. (Indeed, that's part of the reason to use Habitat.)

"Binlinking" will _remove_ some of the environmental correctness guarantees that Habitat gives you, as you will be executing the software in the environment of your current shell.  By way of example, consider running the `core/coreutils` `env` command directly and via `hab pkg exec`. 

```
[10][default:/src:1]# /hab/pkgs/core/coreutils/8.30/20190115012313/bin/env
... elided for space
PATH=/hab/pkgs/core/hab-plan-build/0.78.0/20190313122208/bin:/hab/pkgs/core/diffutils/3.6/20190115013221/bin:/hab/pkgs/core/less/530/20190115013008/bin:/hab/pkgs/core/make/4.2.1/20190115013626/bin:/hab/pkgs/core/mg/20180408/20190115015655/bin:/hab/pkgs/core/util-linux/2.32/20190115013746/bin:/hab/pkgs/core/vim/8.1.0577/20190115015449/bin:/hab/pkgs/core/ncurses/6.1/20190115012027/bin:/hab/pkgs/core/acl/2.2.53/20190115012136/bin:/hab/pkgs/core/attr/2.4.48/20190115012129/bin:/hab/pkgs/core/bash/4.4.19/20190115012619/bin:/hab/pkgs/core/binutils/2.31.1/20190115003743/bin:/hab/pkgs/core/bzip2/1.0.6/20190115011950/bin:/hab/pkgs/core/coreutils/8.30/20190115012313/bin:/hab/pkgs/core/file/5.34/20190115003731/bin:/hab/pkgs/core/findutils/4.6.0/20190115013303/bin:/hab/pkgs/core/gawk/4.2.1/20190115012752/bin:/hab/pkgs/core/glibc/2.27/20190115002733/bin:/hab/pkgs/core/grep/3.1/20190115012541/bin:/hab/pkgs/core/gzip/1.9/20190115013612/bin:/hab/pkgs/core/hab/0.78.0/20190313115951/bin:/hab/pkgs/core/libcap/2.25/20190115012150/bin:/hab/pkgs/core/openssl-fips/2.0.16/20190115014207/bin:/hab/pkgs/core/openssl/1.0.2r/20190305210149/bin:/hab/pkgs/core/pcre/8.42/20190115012526/bin:/hab/pkgs/core/rq/0.10.4/20190115014520/bin:/hab/pkgs/core/sed/4.5/20190115012152/bin:/hab/pkgs/core/tar/1.30/20190115012709/bin:/hab/pkgs/core/unzip/6.0/20190115014516/bin:/hab/pkgs/core/wget/1.19.5/20190305211748/bin:/hab/pkgs/core/xz/5.2.4/20190115013348/bin:/bin

[11][default:/src:0]# hab pkg exec core/coreutils env
... elided for space
PATH=/hab/pkgs/core/coreutils/8.30/20190115012313/bin:/hab/pkgs/core/glibc/2.27/20190115002733/bin:/hab/pkgs/core/acl/2.2.53/20190115012136/bin:/hab/pkgs/core/attr/2.4.48/20190115012129/bin:/hab/pkgs/core/libcap/2.25/20190115012150/bin
```

If we take a look at something like `core/ansible` which uses `push_runtime_env` to set PYTHON_PATH:
```
[14][default:/src:0]#  hab pkg exec core/ansible ansible --version
ansible 2.7.6
  config file = /hab/pkgs/core/ansible/2.7.6/20190305222954/etc/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /hab/pkgs/core/ansible/2.7.6/20190305222954/lib/python2.7/site-packages/ansible-2.7.6-py2.7.egg/ansible
  executable location = /hab/pkgs/core/ansible/2.7.6/20190305222954/bin/ansible
  python version = 2.7.15 (default, Mar  5 2019, 21:30:01) [GCC 8.2.0]

[16][default:/src:1]# hab pkg binlink core/ansible ansible
» Binlinking ansible from core/ansible into /bin
★ Binlinked ansible from core/ansible/2.7.6/20190305222954 to /bin/ansible
[17][default:/src:1]# ansible --version
Traceback (most recent call last):
  File "/bin/ansible", line 4, in <module>
    __import__('pkg_resources').run_script('ansible==2.7.6', 'ansible')
  File "/hab/pkgs/core/python2/2.7.15/20190305212819/lib/python2.7/site-packages/pkg_resources/__init__.py", line 3088, in <module>
    @_call_aside
  File "/hab/pkgs/core/python2/2.7.15/20190305212819/lib/python2.7/site-packages/pkg_resources/__init__.py", line 3072, in _call_aside
    f(*args, **kwargs)
  File "/hab/pkgs/core/python2/2.7.15/20190305212819/lib/python2.7/site-packages/pkg_resources/__init__.py", line 3101, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/hab/pkgs/core/python2/2.7.15/20190305212819/lib/python2.7/site-packages/pkg_resources/__init__.py", line 574, in _build_master
    ws.require(__requires__)
  File "/hab/pkgs/core/python2/2.7.15/20190305212819/lib/python2.7/site-packages/pkg_resources/__init__.py", line 892, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/hab/pkgs/core/python2/2.7.15/20190305212819/lib/python2.7/site-packages/pkg_resources/__init__.py", line 778, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'ansible==2.7.6' distribution was not found and is required by the application
```

I do agree that we should exercise the command, but only in a way to determine that we have something on disk that is executable. More in depth testing would ideally be provided by the upstream software and can be run in `do_check()`.

Reply:
I don't follow this discussion 100%, but I do not understand an advantage in having Habitat binlink then making the test simply call `command`.
The use of `hab pkg exec` was intentionally to give some assurance that the exe being called and tested is definitely the last exe that was built and not another copy of `sed`.

> I would consider it essential to test that the binary you are calling is exercising the binlink, because otherwise you may have a problem with the HAB_BINLINK_DIR environment variable or pkg_bin_dirs directory in the plan.

I could make the test call `/bin/sed` directly but add an additional test case to ensure that `ls -l /bin/sed` confirms that it is linked to the last build.
I'm suspicious about just calling `sed` because on my local setup `which sed` indicates that `sed` points to an old build; even after using `hab pkg install -b my_sed.hart --force`

Reply:
`sed` appears to always point to an old build because it's a dependency of studio itself and so always appears earlier on the path.

Looking into how the binlinking works; the source of the binlink is determined by `/hab/pkgs/core/sed/4.5/[BUILD STAMP]/PATH` and the target of the link is determined by hab cli and not the plan - therefore there doesn't seem to be additional value in having the test call the binlink

Reply:
@echohack given the discussion above, do you still advise changes to this PR, or can this thread be marked Resolved?

Reply:
Hi @james-stocks.  I'm happy for this PR to be approved and merged as the testing here fulfils the aim and standard of the [test design](https://github.com/habitat-sh/core-plans/pull/2539)